packages/@react-aria/table/src/useTableRow.ts

```diff
@@ -13,21 +13,67 @@
 import {FocusableElement} from '@react-types/shared';
 import {getRowLabelledBy} from './utils';
 import {GridRowAria, GridRowProps, useGridRow} from '@react-aria/grid';
-import {RefObject} from 'react';
+import {HTMLAttributes, RefObject} from 'react';
+import {mergeProps} from '@react-aria/utils';
 import {TableCollection} from '@react-types/table';
-import {TableState} from '@react-stately/table';
+import {TableState, TreeGridState} from '@react-stately/table';
+import {useLocale} from '@react-aria/i18n';
+
+const EXPANSION_KEYS = {
+  expand: {
+    ltr: 'ArrowRight',
+    rtl: 'ArrowLeft'
+  },
+  'collapse': {
+    ltr: 'ArrowLeft',
+    rtl: 'ArrowRight'
+  }
+};
 
 /**
  * Provides the behavior and accessibility implementation for a row in a table.
  * @param props - Props for the row.
  * @param state - State of the table, as returned by `useTableState`.
  */
-export function useTableRow<T>(props: GridRowProps<T>, state: TableState<T>, ref: RefObject<FocusableElement>): GridRowAria {
-  let {node} = props;
+export function useTableRow<T>(props: GridRowProps<T>, state: TableState<T> | TreeGridState<T>, ref: RefObject<FocusableElement>): GridRowAria {
+  let {node, isVirtualized} = props;
   let {rowProps, ...states} = useGridRow<T, TableCollection<T>, TableState<T>>(props, state, ref);
+  let {direction} = useLocale();
+
+  if (isVirtualized && !('expandedKeys' in state)) {
+    rowProps['aria-rowindex'] = node.index + 1 + state.collection.headerRows.length; // aria-rowindex is 1 based
+  } else {
+    delete rowProps['aria-rowindex'];
+  }
+
+  let treeGridRowProps: HTMLAttributes<HTMLElement> = {};
+  if ('expandedKeys' in state) {
+    let treeNode = state.keyMap.get(node.key);
+    if (treeNode != null) {
+      let hasChildRows = treeNode.props?.childItems || treeNode.props?.children?.length > state.userColumnCount;
```

Comment:
I can give this a shot. Was a bit wary of doing this since CollectionBuild has some logic for hasChildNodes and not sure if RAC collections could/would make this distinction either compatibility wise

Reply:
Tried it out, setting `hasChildNodes` in Row to false breaks rendering of the Row entirely since CollectionBuilder [early returns](https://github.com/adobe/react-spectrum/blob/b4e1f1cb7d37ec85be729e2cd68d20df6f395c03/packages/%40react-stately/collections/src/CollectionBuilder.ts#L182-L184) in the childNodes generator if that property is false